File: 2021/java/745.prefix-suffix-search.java

/**
Take "apple" as an example, we will insert add "apple#apple", "pple#apple", "ple#apple", "le#apple", "e#apple", "#apple" into the Trie Tree.
If the query is: prefix = "app", suffix = "le", we can find it by querying our trie for
"le#app".
*/
class WordFilter {
    TrieNode trie;
    public WordFilter(String[] words) {
        trie = new TrieNode();
        for (int i = 0; i < words.length; i++) {
            String word = words[i];
            // insert word 
            for (int j = 0; j <= word.length(); j++) {
                insert(word.substring(j) + '#' + word, i);
            }
        }
    }
    
    private void insert(String word, int index) {
        TrieNode curr = trie;
        // System.out.println(word + ", index= " + index);
        for (char ch : word.toCharArray()) {
            if (!curr.children.containsKey(ch)) {
                curr.children.put(ch, new TrieNode());
            }
            curr = curr.children.get(ch);
            curr.index = index;
        }
    }
    
    private int search(String prefix) {
        TrieNode curr = trie;
        for (char ch : prefix.toCharArray()) {
            if (!curr.children.containsKey(ch)) return -1;
            curr = curr.children.get(ch);
        }
        return curr.index;
    }
    public int f(String prefix, String suffix) {
        return search(suffix + '#' + prefix);
    }
}
class TrieNode {
    Map<Character, TrieNode> children;
    int index;
    public TrieNode() {
        children = new HashMap<>();
        index = 0;
    }
}

File: 2021/python/745.prefix-suffix-search.py
# Create a Trie. Each TrieNode has idx dictionary to store the index of the word.
class TrieNode:
    def __init__(self):
        self.nodes = defaultdict(TrieNode)
        self.isWordEnd = False
        self.idx = {}
    
class Trie:
    def __init__(self):
        self.root = TrieNode()
    
	# Insert the word into trie
    def insert(self, word, idx):
        curr = self.root
        for char in word:
            curr.idx[char] = idx
            curr = curr.nodes[char]
        curr.isWordEnd = True
     
    # Search for the prefix and keep updating the idx  
    def prefixSearch(self, prefix):
        curr = self.root
        prefixIdx = -1
        for char in prefix:
            if char not in curr.nodes:
                break
            else:
                prefixIdx = curr.idx[char]
                curr = curr.nodes[char]
        return prefixIdx
        
class WordFilter:
    def __init__(self, words: List[str]):
        self.wordMap = {}
        self.trie = Trie()
        
		# For each word create string using the suffix
		# apple => #apple, e#apple, le#apple, ple#apple, pple#apple, apple#apple 
        for i in range(len(words)):
            word = words[i]
            self.trie.insert('#'+word, i)
            for j in range(1, len(word)+1):
                newWord = word[-j:] + '#' + word
                self.trie.insert(newWord, i)
            
        
    def f(self, prefix: str, suffix: str) -> int:
		# generate the prefix to search
        searchKey = suffix + '#' + prefix
        return self.trie.prefixSearch(searchKey)
